File: jianzhioffer/firstround/jianzhi_45.py
# 选择排序
# class Solution(object):
#     def minNumber(self, nums):
#         """
#         :type nums: List[int]
#         :rtype: str
#         """
#         if not nums:
#             return ""
#         nums = [str(num) for num in nums]
#
#         for i in range(len(nums)-1):
#             min_idx = i
#             for j in range(i+1, len(nums)):
#                 if nums[j] + nums[min_idx] < nums[min_idx] + nums[j]:
#                     min_idx = j
#             nums[i], nums[min_idx] = nums[min_idx], nums[i]
#         return ''.join(nums)

# 快速排序
class Solution(object):
    def minNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: str
        """
        if not nums:
            return ""

        def fast_sort(left, right):
            if left >= right:
                return
            i, j = left, right
            while i < j:
                while i < j and nums[j] + nums[left] >= nums[left] + nums[j]:
                    j -= 1
                while i < j and nums[i] + nums[left] <= nums[left] + nums[i]:
                    i += 1
                nums[i], nums[j] = nums[j], nums[i]
            nums[i], nums[left] = nums[left], nums[i]
            fast_sort(left, i-1)
            fast_sort(i+1, right)

        nums = [str(num) for num in nums]
        fast_sort(0, len(nums)-1)
        return ''.join(nums)

if __name__ == '__main__':
    nums = [1,4,7,2,5,8,0,3,6,9]
    nums.sort()
    obj = Solution()
    res = obj.minNumber(nums)
    print(res)

File: jianzhioffer/secondround/jianzhi_45.cpp
#include <string>
#include <vector>
using namespace std;

//快速排序，时间复杂度O(NlogN)，空间复杂度O(N)
//能使用排序解题的证明见leetcode题解
class Solution
{
public:
    Solution();
    ~Solution();
    string minNumber(vector<int>& nums);


};

Solution::Solution()
{

}

Solution::~Solution()
{

}

int Partition(vector<int>& nums, int left, int right)
{
    int pivot = nums[left];
    while (left<right)
    {
        while(left<right && ((to_string(nums[right])+to_string(pivot))>=(to_string(pivot)+to_string(nums[right])))) right--;
        nums[left] = nums[right];
        while(left<right && ((to_string(nums[left])+to_string(pivot))<=(to_string(pivot)+to_string(nums[left])))) left++;
        nums[right] = nums[left];
    }
    nums[left] = pivot;
    return left;   
}

void QuickSort(vector<int>& nums, int left, int right)
{
    if(left>=right) return;
    int pivot_position = Partition(nums, left, right);
    QuickSort(nums, left, pivot_position-1);
    QuickSort(nums, pivot_position+1, right);
}

string Solution::minNumber(vector<int>& nums)
{
    if(nums.empty()) return "";
    QuickSort(nums, 0, nums.size()-1);
    string res = "";
    for (int i = 0; i < nums.size(); i++)
    {
        res += to_string(nums[i]);
    }
    return res;
}


int main(int argc, char const *argv[])
{
    vector<int> nums = {3,30,34,5,9};
    Solution obj;
    string result = obj.minNumber(nums);
    return 0;
}
